File: src/main/java/one/oktw/sponge/internal/galaxy/Planet.java
package one.oktw.sponge.internal.galaxy;

import com.mongodb.client.MongoDatabase;
import one.oktw.sponge.Main;
import org.bson.Document;
import org.spongepowered.api.Server;
import org.spongepowered.api.Sponge;
import org.spongepowered.api.world.World;
import org.spongepowered.api.world.storage.WorldProperties;

import java.util.Optional;
import java.util.UUID;

import static com.mongodb.client.model.Filters.eq;

public class Planet {
    private Server server = Sponge.getServer();
    private Main main = Main.getMain();
    private PlanetManager planetManager = main.getPlanetManager();
    private MongoDatabase database = main.getDatabaseManager().getDatabase();
    private Document planet;

    Planet(UUID uuid) {
        this.planet = database.getCollection("World").find(eq("UUID", uuid)).first();
    }

    public UUID getUniqueId() {
        return (UUID) planet.get("UUID");
    }

    public String getName() {
        return planet.getString("Name");
    }

    public int getSize() {
        return planet.getInteger("Size");
    }

    public Optional<World> getWorld() {
        return planetManager.loadWorld(getUniqueId());
    }

    public Optional<WorldProperties> getWorldProp() {
        return server.getWorldProperties(getUniqueId());
    }

    public int setSize(int size) {
        int originSize = getSize();
        WorldProperties properties = getWorldProp().get();

        database.getCollection("World").findOneAndUpdate(
                eq("UUID", getUniqueId()),
                new Document("$set", new Document("Size", size))
        );

        properties.setWorldBorderTargetDiameter(size * 16);
        server.saveWorldProperties(properties);

        return size - originSize;
    }

    public void setSecurity(int level) {
        database.getCollection("world").findOneAndUpdate(
                eq("UUID", getUniqueId()),
                new Document("$set", new Document("Security", level))
        );
    }
}

File: src/main/kotlin/one/oktw/galaxy/internal/Planet.kt
package one.oktw.galaxy.internal

import com.mongodb.client.model.Filters.eq
import com.mongodb.client.model.Updates.set
import one.oktw.galaxy.Main
import one.oktw.galaxy.internal.PlanetManager.Companion.loadPlanet
import org.bson.Document
import org.spongepowered.api.Sponge
import org.spongepowered.api.world.World
import org.spongepowered.api.world.storage.WorldProperties
import java.util.*

class Planet internal constructor(uuid: UUID) {
    private val server = Sponge.getServer()
    private val database = Main.databaseManager.database
    private val planets = database.getCollection("Planet")
    private val planet: Document

    init {
        this.planet = planets.find(eq("UUID", uuid)).first()
    }

    val uniqueId: UUID
        get() = planet["UUID"] as UUID

    var name: String
        get() = planet.getString("Name")
        set(name) {
            planets.findOneAndUpdate(eq("UUID", uniqueId), set("name", name))
        }

    var size: Int
        get() = planet.getInteger("Size")
        set(size) {
            val properties = worldProp.get()

            properties.worldBorderTargetDiameter = (size * 16).toDouble()
            if (server.saveWorldProperties(properties)) {
                planets.findOneAndUpdate(eq("UUID", uniqueId), set("Size", size))
            }
        }

    var security: SecurityLevel
        get() = SecurityLevel.fromInt(planet.getInteger("Security"))
        set(securityLevel) {
            planets.findOneAndUpdate(eq("UUID", uniqueId), set("Security", securityLevel.level))
        }

    val world: Optional<World>
        get() = loadPlanet(uniqueId)

    val worldProp: Optional<WorldProperties>
        get() = server.getWorldProperties(uniqueId)
}
